src/tool_writeout_json.c

```diff
@@ -0,0 +1,201 @@
+  {"http_version", JSON_VERSION, CURLINFO_HTTP_VERSION},
+  {"scheme", JSON_STRING, CURLINFO_SCHEME},
+  {NULL, JSON_NONE, CURLINFO_NONE}
+};
+
+static const char *http_version[] = {
+  "0",   /* CURL_HTTP_VERSION_NONE */
+  "1",   /* CURL_HTTP_VERSION_1_0 */
+  "1.1", /* CURL_HTTP_VERSION_1_1 */
+  "2"    /* CURL_HTTP_VERSION_2 */
+  "3"    /* CURL_HTTP_VERSION_3 */
+};
+
+static void json_escape(FILE *stream, const char *in)
+{
+  const char *i = in;
+  const char *in_end = in + strlen(in);
+
+  for(; i < in_end; i++) {
+    switch(*i) {
+    case '\\':
+      fputs("\\\\", stream);
+      break;
+    case '\"':
+      fputs("\\\"", stream);
+      break;
+    case '\b':
+      fputs("\\b", stream);
+      break;
+    case '\f':
+      fputs("\\f", stream);
+      break;
+    case '\n':
+      fputs("\\n", stream);
+      break;
+    case '\r':
+      fputs("\\r", stream);
+      break;
+    case '\t':
+      fputs("\\t", stream);
+      break;
+    default:
+      if (*i < 32) {
+        fprintf(stream, "u%04x", *i);
+      }
+      else {
+        fputc(*i, stream);
+      }
+      break;
+    };
+  }
+}
+
+
+void ourWriteOutJSON(CURL *curl, struct OutStruct *outs, FILE *stream)
+{
+  char *stringp = NULL;
+  long longinfo;
+  double doubleinfo;
+  int i;
+
+#ifdef HAVE_SETLOCALE
+  /* to produce valid JSON: disable any locale conversion of numbers */
+  const char *current_locale = setlocale(LC_NUMERIC, NULL);
+  setlocale(LC_NUMERIC, "POSIX");
+#endif
+
+  fputs("{", stream);
+  for(i = 0; mappings[i].key != NULL; i++) {
+    const char *k = mappings[i].key;
+    CURLINFO cinfo = mappings[i].cinfo;
+    switch(mappings[i].type) {
+    case JSON_STRING:
+      if((CURLE_OK == curl_easy_getinfo(curl, cinfo, &stringp)) && stringp) {
+        fprintf(stream, "\"%s\":\"", k);
+        json_escape(stream, stringp);
+        fprintf(stream, "\",");
+      }
+      break;
+    case JSON_LONG:
+      longinfo = 0;
+      if(CURLE_OK == curl_easy_getinfo(curl, cinfo, &longinfo)) {
+        fprintf(stream, "\"%s\":%ld,", k, longinfo);
+      }
+      break;
+    case JSON_DOUBLE:
+      if(CURLE_OK == curl_easy_getinfo(curl, cinfo, &doubleinfo)) {
+        fprintf(stream, "\"%s\":%.6f,", k, doubleinfo);
+      }
+      break;
+    case JSON_FILENAME:
+      if(outs->filename) {
+        fprintf(stream, "\"%s\":\"", k);
+        json_escape(stream, outs->filename);
+        fprintf(stream, "\",");
+
+      }
+      else {
+        fprintf(stream, "\"%s\":null,", k);
+      }
+      break;
+    case JSON_VERSION:
+      if(CURLE_OK == curl_easy_getinfo(curl, cinfo, &longinfo) &&
+         (longinfo >= 0) &&
+         (longinfo < (long)(sizeof(http_version)/sizeof(char *)))) {
+        fprintf(stream, "\"%s\":\"%s\",", k, http_version[longinfo]);
+      }
+      break;
+    default:
+      break;
+    }
+  }
+  fprintf(stream, "\"curl_version\":\"%s\"}", curl_version());
+
+#ifdef HAVE_SETLOCALE
+  setlocale(LC_NUMERIC, current_locale);
```

Comment:
This causes a valgrind error:
~~~
==1709491==    at 0x50006A0: __strcmp_sse2_unaligned (strcmp-sse2-unaligned.S:24)
==1709491==    by 0x4F9E037: setlocale (setlocale.c:238)
==1709491==    by 0x143995: ourWriteOutJSON (tool_writeout_json.c:199)
~~~
The reason for this is that the `current_locale` pointer that is returned by the first `setlocale()` call in this function (and returns to some static content) is ruin by the *second* call to `setlocale()` so this third call apparently potentially accesses invalid memory!

Reply:
The valgrind trigger is easily reproduced with: `valgrind ./src/curl localhost --write-out '%{json}'`

Reply:
will look into this.

Reply:
I think you can solve it by not calling it with a NULL first, as it will return the only one even when you set it to "POSIX". But perhaps you don't even need to change the locale setting at all if you avoid the double values...

Reply:
for downloaded-bytes for sure. but for time in seconds?

Reply:
The `_T` versions of the time values return time as microseconds so you could easily printf them "manually" %u.%06u - style. Thus avoid any system time formatting.

Reply:
works.